addons/zylann.hterrain/native/src/quad_tree_lod.h

```diff
@@ -0,0 +1,122 @@
+#ifndef QUAD_TREE_LOD_H
+#define QUAD_TREE_LOD_H
+
+#include <Godot.hpp>
+#include <FuncRef.hpp>
+
+namespace godot {
+
+class QuadTreeLod : public Reference {
+    GODOT_CLASS(QuadTreeLod, Reference)
+
+    class Quad {
+    private:
+        Quad **children = nullptr;
```

Comment:
This is a dynamically-allocated array for fixed amount of items, which itself contains dynamically-allocated quads. In GDScript there was no other choice, but in C++ it is possible to rather use a single vector of Quads like a pool where every pack of 4 is used for children. An index to the first child can be used instead of a pointer, and the boolean can be removed by interpreting a negative index as having no children. This way we go from 16-byte structs scattered on the heap to 4-byte structs close together in memory (it has its own tricky parts, but I've done that with an octree: https://github.com/Zylann/godot_voxel/blob/master/terrain/lod_octree.h). It might bring minor improvement in 2D but it could help. I dont think this causes your slowdown though.